### ISSUE TITLE
BUG: approx_derivative fails for complex-valued functions

### ISSUE THREAD

**Issue by KeithWM** [scipy/scipy#18772]
### Describe your issue.

I'm using `scipy.optimizer.check_grad` to test the implementation of a Jacobian for a complex-valued function. Unfortunatelt, the approximated numerical derivatve in `_dense_difference` (called from `approx_derivative`) casts the Jacobian to real by initializing `J_transposed = np.empty((n, m))`. I think this could be easily fixed by including a `dtype=f0.dtype` keyword argument to the `empty` call.

### Reproducing Code Example

```python
import scipy
scipy.optimize._numdiff.approx_derivative(lambda x: 1.j * x, 1)
# should return 1j, return 0
```

### Error message

```shell
\optimize\_numdiff.py:598: ComplexWarning: Casting complex values to real discards the imaginary part
  J_transposed[i] = df / dx
```

### SciPy/NumPy/Python version and system information

```shell
scipy version 1.10.1
numpy 1.24.3
python 3.11.3
```

**Comment by andyfaff** [comment]
A few points here. Firstly, `approx_derivative` is currently a private function, so should not be relied upon to be unchanging by external users. Secondly, the docstring clearly states that `approx_derivative` works in the real plane:

> R^n to R^m

**Comment by mdhaber** [comment]
Agreed @andyfaff. I think it's fine to close this. The request for public numerical differentiation is noted elsewhere (e.g. gh-2035 , gh-7457, gh-17059) , so this does not need to remain open as an enhancement request. Note that I will be opening a PR for numerical differentiation of elementwise scalar-in, scalar-out functions shortly (based on https://github.com/mdhaber/scipy/pull/106; see also https://github.com/mdhaber/jacobi/pull/1). I can consider supporting complex functions, and I certainly plan to extend it to the vector-in, vector-out case.

**Comment by KeithWM** [comment]
I would like to point out that even if `approx_derivative` is a private function, its narrow implementation is causing `check_grad` to fail. I'm a bit disappointed at the decision to close this issue so readily, while the solution is so simple.

Anyway, I will create a pull request implementing the fix. Perhaps that will also illustrate just how trivial the proposed change is.

**Comment by mdhaber** [comment]
There are a lot of open issues, so we have to triage, and one of those steps is whether the function is public or not. 
`check_grad` is public, so something that fixes that (or documents that complex-valued functions are not supported) sounds good.

**Comment by andyfaff** [comment]
> while the solution is so simple

That statement is untrue.

- If you're differentiating a complex function you should be creating steps in the imaginary axis as well. That in itself would require a lot of work.
-  There are many different directions one could approach z_0. Is the derivative independent of the direction one approaches z_0 with the finite step? (What about Cauchy Riemann conditions?)
-  because the value of the derivative (of a complex function) could depend on which direction the step is taken in one would need partial derivatives. How would that interact with the real machinery which is less complex?
- how would the approx_derivative know if you were differentiating a complex function or not?

**Comment by KeithWM** [comment]
I'm talking about a complex-valued function with real argument(s). Otherwise it is indeed a completely different story. 

I've made the necessary modification to the code and written a test for it, but setting up the scipy environment.yml is proving a lot harder than I thought it would be; I run into conflicting dependencies on my system. Can the tests be triggered by a github action on creating a PR?
